scripts/generate-embeddings.ts

```diff
@@ -0,0 +1,426 @@
+// script taken from:
+// https://github.com/supabase-community/nextjs-openai-doc-search/blob/main/lib/generate-embeddings.ts
+
+import { createClient } from '@supabase/supabase-js';
+import { createHash } from 'crypto';
+import dotenv from 'dotenv';
+import { readdir, readFile, stat } from 'fs/promises';
+import GithubSlugger from 'github-slugger';
+import { fromMarkdown } from 'mdast-util-from-markdown';
+import { toMarkdown } from 'mdast-util-to-markdown';
+import { toString } from 'mdast-util-to-string';
+import 'openai';
+import { Configuration, OpenAIApi } from 'openai';
+import { basename, dirname, join } from 'path';
+import { u } from 'unist-builder';
+import { inspect } from 'util';
+import yargs from 'yargs';
+
+dotenv.config();
+
+/**
+ * Splits a `mdast` tree into multiple trees based on
+ * a predicate function. Will include the splitting node
+ * at the beginning of each tree.
+ *
+ * Useful to split a markdown file into smaller sections.
+ */
+function splitTreeBy(tree: any, predicate: (node: any) => boolean) {
+  return tree.children.reduce((trees: any, node: any) => {
+    const [lastTree] = trees.slice(-1);
+
+    if (!lastTree || predicate(node)) {
+      const tree = u('root', [node]);
+      return trees.concat(tree);
+    }
+
+    lastTree.children.push(node);
+    return trees;
+  }, []);
+}
+
+type Section = {
+  content: string;
+  heading?: string;
+  slug?: string;
+};
+
+type ProcessedMdx = {
+  checksum: string;
+  sections: Section[];
+};
+
+/**
+ * Processes MD content for search indexing.
+ * It extracts metadata and splits it into sub-sections based on criteria.
+ */
+function processMdxForSearch(content: string): ProcessedMdx {
+  const checksum = createHash('sha256').update(content).digest('base64');
+
+  const mdTree = fromMarkdown(content, {});
+
+  if (!mdTree) {
+    return {
+      checksum,
+      sections: [],
+    };
+  }
+
+  const sectionTrees = splitTreeBy(mdTree, (node) => node.type === 'heading');
+
+  const slugger = new GithubSlugger();
+
+  const sections = sectionTrees.map((tree: any) => {
+    const [firstNode] = tree.children;
+
+    const heading =
+      firstNode.type === 'heading' ? toString(firstNode) : undefined;
+    const slug = heading ? slugger.slug(heading) : undefined;
+
+    return {
+      content: toMarkdown(tree),
+      heading,
+      slug,
+    };
+  });
+
+  return {
+    checksum,
+    sections,
+  };
+}
+
+type WalkEntry = {
+  path: string;
+  parentPath?: string;
+};
+
+async function walk(dir: string, parentPath?: string): Promise<WalkEntry[]> {
```

Comment:
We already have a map of the source we want to use in `map.json`, as well as information about the executors and generators with `packages.json`. We also have already made, pre-digested data by type in the `docs/generated/manifests` folder. Did you look into it, was it not useful?

Reply:
I did not look into it, no! I will, though, now :D

Reply:
Done! :D